### PR TITLE
Refactor Consul Syncer into new ServiceClient

### LINKED ISSUE #2478: Consul Service Advertisement Refactor
*There have been enough issues related to this work I wanted to create an issue to link them all *
together.*

Branch: https://github.com/hashicorp/nomad/tree/f-consul-refactor

The existing (0.5.5 and before) Consul service advertisement code has the following properties:

* Runs in the task executor so checks are updated even if the Nomad agent is down.
* Designed as a reconciliation loop to ensure Consul is consistent with Nomad (eg reaps orphaned Consul entries periodically)

This has the following issues:

* Reconciliation loop is expensive and must read all services out of Consul periodically even if nothing has changed.
* Since each task executor runs its own reconciliation loop this resource utilization gets far worse for users with many tasks. See #1995 
* Reconciliation logic is complex and error prone. See #2474

The goal of the Consul refactor is to address these issues by doing the following:

* Add/remove services/checks as appropriate. No periodic reconciliation.
* Move all Consul interaction into the Nomad client agent - not the task executor.
  * The one downside to this is that if the Nomad agent is down for longer than a script check's interval, the script check will not be run and the service's health in Consul will not be accurate.

### PR DESCRIPTION
**Pull request by schmichael**
Fixes #2478 #2474 #1995 #2294

The new client only handles agent and task service advertisement. Server
discovery is mostly unchanged.

The Nomad client agent now handles all Consul operations instead of the
executor handling task related operations. When upgrading from an
earlier version of Nomad existing executors will be told to deregister
from Consul so that the Nomad agent can re-register the task's services
and checks.

Drivers - other than qemu - now support an Exec method for executing
abritrary commands in a task's environment. This is used to implement
script checks.

Interfaces are used extensively to avoid interacting with Consul in
tests that don't assert any Consul related behavior.

### CODE REVIEW

File: command/agent/consul/client.go [hunk 1]
```diff
@@ -0,0 +1,636 @@
...
+}
+
+// NewServiceClient creates a new Consul ServiceClient from an existing Consul API
+// Client and logger.
+func NewServiceClient(consulClient AgentAPI, logger *log.Logger) *ServiceClient {
+	return &ServiceClient{
+		client:         consulClient,
+		logger:         logger,
+		retryInterval:  defaultSyncInterval, //TODO what should this default to?!
+		syncInterval:   defaultSyncInterval,
```
Comment (schmichael) [inline command/agent/consul/client.go]: unused?

File: command/agent/consul/client.go [hunk 2]
```diff
@@ -0,0 +1,636 @@
...
+		select {
+		case <-c.syncCh:
+			timer.Reset(0)
+		case <-timer.C:
+			if err := c.sync(); err != nil {
+				if lastOk {
+					lastOk = false
+					c.logger.Printf("[WARN] consul: failed to update services in Consul: %v", err)
+				}
+				//TODO Log? and jitter/backoff
```
Comment (schmichael) [inline command/agent/consul/client.go]: exp backoff from 1...30s

File: command/agent/consul/client.go [hunk 3]
```diff
@@ -0,0 +1,636 @@
...
+		if err = c.client.ServiceDeregister(id); err != nil {
+			goto ERROR
+		}
+		delete(deregServices, id)
+	}
+
+	c.logger.Printf("[DEBUG] consul: registered %d services / %d checks; deregisterd %d services / %d checks", regServiceN, regCheckN, deregServiceN, deregCheckN)
+	return nil
+
+	//TODO Labels and gotos are nasty; move to a function?
```
Comment (schmichael) [inline command/agent/consul/client.go]: add helper functions to get a snapshot of the maps and another function to merge outstanding functions back in

File: command/agent/consul/script.go [hunk 4]
```diff
@@ -0,0 +1,136 @@
...
+			err = s.agent.UpdateTTL(s.id, string(output), state)
+			select {
+			case <-ctx.Done():
+				// check has been removed; don't report errors
+				return
+			default:
+			}
+
+			if err != nil {
+				//FIXME Backoff? Retry faster?
```
Comment (schmichael) [inline command/agent/consul/script.go]: remove this

File: client/task_runner.go [hunk 5]
```diff
@@ -289,6 +292,19 @@ func (r *TaskRunner) RestoreState() error {
 				r.task.Name, r.alloc.ID, err)
 			return nil
 		}
+
+		//FIXME is there a better place to do this? used to be in executor
+		// Prepare services
+		interpolateServices(r.getTaskEnv(), r.task)
```
Comment (schmichael) [inline client/task_runner.go]: Move this

File: command/agent/consul/script_test.go [hunk 6]
```diff
@@ -0,0 +1,165 @@
...
+import (
+	"context"
+	"os/exec"
+	"testing"
+	"time"
+
+	"github.com/hashicorp/consul/api"
+	"github.com/hashicorp/nomad/nomad/structs"
+)
+
```
Comment (dadgar) [inline command/agent/consul/script_test.go]: Where are the normal case tests:
* Getting correct exit code
* Getting output
* Getting exec errors

File: client/driver/utils.go [hunk 7]
```diff
@@ -181,3 +181,36 @@ func getExecutorUser(task *structs.Task) string {
 	}
 	return task.User
 }
+
+// execChroot executes cmd with args inside chroot if set and returns the
+// output, exit code, and error. If chroot is an empty string the command is
+// executed on the host.
+func execChroot(ctx context.Context, chroot, name string, args []string) ([]byte, int, error) {
```
Comment (schmichael) [inline client/driver/utils.go]: Added #2554 to handle this in a followup PR.

File: client/driver/raw_exec.go [hunk 8]
```diff
@@ -250,6 +245,10 @@ func (h *rawExecHandle) Update(task *structs.Task) error {
 	return nil
 }
 
+func (h *rawExecHandle) Exec(ctx context.Context, cmd string, args []string) ([]byte, int, error) {
+	return execChroot(ctx, "", cmd, args)
```
Comment (schmichael) [inline client/driver/raw_exec.go]: Handled in #2554

File: client/driver/java.go [hunk 9]
```diff
@@ -390,6 +388,10 @@ func (h *javaHandle) Update(task *structs.Task) error {
 	return nil
 }
 
+func (h *javaHandle) Exec(ctx context.Context, cmd string, args []string) ([]byte, int, error) {
+	return execChroot(ctx, h.taskDir, cmd, args)
```
Comment (schmichael) [inline client/driver/java.go]: Handled in #2554

File: command/agent/consul/int_test.go [hunk 10]
```diff
@@ -0,0 +1,228 @@
...
+
+	// Kill the task
+	tr.Kill("", "", false)
+
+	select {
+	case <-tr.WaitCh():
+	case <-time.After(10 * time.Second):
+		t.Fatalf("timed out waiting for Run() to exit")
+	}
+
```
Comment (schmichael) [inline command/agent/consul/int_test.go]: `tr.Kill` a few lines up calls `tr.Destroy`

File: command/agent/consul/int_test.go [hunk 11]
```diff
@@ -0,0 +1,228 @@
...
+	defer serviceClient.Shutdown() // just-in-case cleanup
+	consulRan := make(chan struct{})
+	go func() {
+		serviceClient.Run()
+		close(consulRan)
+	}()
+	tr := client.NewTaskRunner(logger, conf, logUpdate, taskDir, alloc, task, vclient, serviceClient)
+	tr.MarkReceived()
+	go tr.Run()
+	defer func() {
```
Comment (schmichael) [inline command/agent/consul/int_test.go]: Ensuring we don't leak the taskrunner goroutine if the test exits early.

File: command/agent/consul/client.go [hunk 12]
```diff
@@ -0,0 +1,633 @@
...
+			Port:    port,
+		}
+		regs[i] = serviceReg
+
+		for _, check := range service.Checks {
+			checkID := createCheckID(id, check)
+			if check.Type == structs.ServiceCheckScript {
+				return fmt.Errorf("service %q contains invalid check: agent checks do not support scripts", service.Name)
+			}
+			checkHost, checkPort := serviceReg.Address, serviceReg.Port
```
Comment (schmichael) [inline command/agent/consul/client.go]: explained in the line below

File: nomad/server.go [hunk 13]
```diff
@@ -63,6 +63,14 @@ const (
 	// raftRemoveGracePeriod is how long we wait to allow a RemovePeer
 	// to replicate to gracefully leave the cluster.
 	raftRemoveGracePeriod = 5 * time.Second
+
+	// defaultConsulDiscoveryInterval is how often to poll Consul for new
+	// servers if there is no leader.
+	defaultConsulDiscoveryInterval time.Duration = 9 * time.Second
+
+	// defaultConsulDiscoveryIntervalRetry is how often to poll Consul for
+	// new servers if there is no leader and the last Consul query failed.
```
Comment (schmichael) [inline nomad/server.go]: Fixed!

File: command/agent/consul/client.go [hunk 14]
```diff
@@ -0,0 +1,678 @@
+package consul
```
Comment (dadgar) [inline command/agent/consul/client.go]: Can you add metrics to this package that output the number of services/checks/script checks being registered and one on each sync failure.

File: command/agent/consul/client.go [hunk 15]
```diff
@@ -0,0 +1,678 @@
...
+			if _, running := c.runningScripts[id]; running {
+				continue
+			}
+			// Not running, start and store the handle
+			c.runningScripts[id] = script.run()
+		}
+	}
+
+	c.logger.Printf("[DEBUG] consul.sync: registered %d services, %d checks; deregistered %d services, %d checks",
+		sreg, creg, sdereg, cdereg)
```
Comment (dadgar) [inline command/agent/consul/client.go]: These would be good metrics too to measure the rate at which things are being modified.

File: command/agent/consul/client.go [hunk 16]
```diff
@@ -0,0 +1,678 @@
...
+			// Existing sevice entry removed
+			ops.deregServices = append(ops.deregServices, existingID)
+			for _, check := range existingSvc.Checks {
+				ops.deregChecks = append(ops.deregChecks, createCheckID(existingID, check))
+			}
+			continue
+		}
+
+		// Service exists and wasn't updated, don't add it later
+		delete(newIDs, existingID)
```
Comment (dadgar) [inline command/agent/consul/client.go]: What if the port changes?

File: command/agent/consul/mock.go [hunk 17]
```diff
@@ -0,0 +1,27 @@
+package consul
+
+import (
+	"log"
+
+	"github.com/hashicorp/consul/api"
```
Comment (dadgar) [inline command/agent/consul/mock.go]: we have been following the habit of <name>_testing.go to put importable stubs/mocks/etc

File: command/agent/consul/script.go [hunk 18]
```diff
@@ -0,0 +1,159 @@
...
+			// Execute check script with timeout
+			execctx, cancel := context.WithTimeout(ctx, s.check.Timeout)
+			output, code, err := s.exec.Exec(execctx, s.check.Command, s.check.Args)
+			switch execctx.Err() {
+			case context.Canceled:
+				// check removed during execution; exit
+				return
+			case context.DeadlineExceeded:
+				// Log deadline exceeded every time, but flip last check to false
+				s.lastCheckOk = false
```
Comment (dadgar) [inline command/agent/consul/script.go]: Won't the `code == 0` in this case?

File: command/agent/consul/script.go [hunk 19]
```diff
@@ -0,0 +1,159 @@
...
+				s.logger.Printf("[WARN] consul.checks: check %q for task %q alloc %q timed out (%s)",
+					s.check.Name, s.taskName, s.allocID, s.check.Timeout)
+			}
+
+			// cleanup context
+			cancel()
+
+			state := api.HealthCritical
+			switch code {
```
Comment (dadgar) [inline command/agent/consul/script.go]: https://www.consul.io/docs/agent/checks.html#check-scripts

Think you want a default case

File: command/agent/consul/script_test.go [hunk 20]
```diff
@@ -0,0 +1,173 @@
...
+	"github.com/hashicorp/nomad/helper/testtask"
+	"github.com/hashicorp/nomad/nomad/structs"
+)
+
+func TestMain(m *testing.M) {
+	if !testtask.Run() {
+		os.Exit(m.Run())
+	}
+}
+
```
Comment (dadgar) [inline command/agent/consul/script_test.go]: Can you test:

* exit code 0
* exit code 1
* exit code 2
* Capturing output
* Unknown command

File: command/agent/consul/client.go (hunk 16 above)
Comment (schmichael) [inline command/agent/consul/client.go]: Totally missed the fact that port changes are excluded from service and check IDs, so great catch. Added a huge test to exercise port changes and did lots of fixing: https://github.com/hashicorp/nomad/pull/2467/commits/970979d8fa31119b425fda35e866b00861f55108

File: command/agent/consul/script.go (hunk 19 above)
Comment (schmichael) [inline command/agent/consul/script.go]: `state` is initialized to the default case (critical/failing)

File: command/agent/consul/script_test.go (hunk 20 above)
Comment (schmichael) [inline command/agent/consul/script_test.go]: Added except "Unknown command" which depends on the Exec implementation to return an error (which is tested): https://github.com/hashicorp/nomad/pull/2467/commits/9c14e8be8cec0fe70262472bd2274303ae04b830

Exec implementations are tested directly for example: https://github.com/hashicorp/nomad/pull/2467/files#diff-1eb2911dd4ed9b612194897284c68251R304

File: command/agent/consul/client.go (hunk 14 above)
Comment (schmichael) [inline command/agent/consul/client.go]: Metrics commit: https://github.com/hashicorp/nomad/pull/2467/commits/1e8b1fcc5179b5a35657dfdf026a4e271900ddff

File: command/agent/consul/client.go [hunk 21]
```diff
@@ -0,0 +1,678 @@
...
+		case ops := <-c.opCh:
+			c.merge(ops)
+		}
+
+		if err := c.sync(); err != nil {
+			if failures == 0 {
+				c.logger.Printf("[WARN] consul.sync: failed to update services in Consul: %v", err)
+			}
+			failures++
+			if !retryTimer.Stop() {
```
Comment (schmichael) [inline command/agent/consul/client.go]: Comment added. I hate the timer API.

File: client/task_runner.go [hunk 22]
```diff
@@ -1061,8 +1072,12 @@ func (r *TaskRunner) run() {
 	}
 }
 
-// cleanup calls Driver.Cleanup when a task is stopping. Errors are logged.
+// cleanup removes Consul entries and calls Driver.Cleanup when a task is
```
Comment (schmichael) [inline client/task_runner.go]: Great catch! Fixed and tested: https://github.com/hashicorp/nomad/pull/2467/commits/9625e19fe86620a552adb547adf4b7018c2c941d

Ensured test failed without the RemoveTask call when restarting.

**Comment by schmichael** [comment]
Ready for another review. Note that #2218 isn't supported for non-agent checks. Will do that in a follow up PR.

File: command/agent/agent.go [hunk 23]
```diff
@@ -680,3 +703,59 @@ func (a *Agent) setupConsul(consulConfig *config.ConsulConfig) error {
 	go a.consulService.Run()
 	return nil
 }
+
+// consulSupportsTLSSkipVerify returns true if Consul supports TLSSkipVerify.
+func consulSupportsTLSSkipVerify(self map[string]map[string]interface{}) bool {
```
Comment (dadgar) [inline command/agent/agent.go]: Use: https://github.com/hashicorp/go-version

File: command/agent/consul/script_test.go [hunk 24]
```diff
@@ -171,3 +227,50 @@ func TestConsulScript_Exec_Shutdown(t *testing.T) {
...
+			}
+			if update.output != expectedOutput {
+				t.Errorf("expected output=%q but found: %q", expectedOutput, update.output)
+			}
+		case <-time.After(3 * time.Second):
+			t.Fatalf("timed out waiting for script check to exec")
+		}
+	}
+
+	// Test exit codes with errors
```
Comment (dadgar) [inline command/agent/consul/script_test.go]: Use go 1.8 table driven test style

File: command/agent/consul/unit_test.go [hunk 25]
```diff
@@ -242,6 +260,188 @@ func TestConsul_ChangeTags(t *testing.T) {
 	}
 }
 
+// TestConsul_ChangePorts asserts that changing the ports on a service updates
+// it in Consul. Since ports are part of the service ID this is a slightly
```
Comment (dadgar) [inline command/agent/consul/unit_test.go]: ports are not part of the service ID

**Comment by dadgar** [comment]
Also fix the diff tests: https://travis-ci.org/hashicorp/nomad/builds/223416717#L5545

File: nomad/structs/structs.go [hunk 26]
```diff
@@ -2074,6 +2074,7 @@ type ServiceCheck struct {
 	Interval      time.Duration // Interval of the check
 	Timeout       time.Duration // Timeout of the response from the check before consul fails the check
 	InitialStatus string        // Initial status of the check
+	TLSSkipVerify bool          // Skip TLS verification when Protocol=https
```
Comment (dadgar) [inline nomad/structs/structs.go]: Add to API package, job spec, tests, diff. 
1) https://github.com/hashicorp/nomad/blob/master/api/tasks.go#L81
2) https://github.com/hashicorp/nomad/blob/master/jobspec/parse.go#L941
3) https://github.com/hashicorp/nomad/blob/master/command/agent/job_endpoint.go#L496

Comment (dadgar) [inline nomad/structs/structs.go]: Also document on both HCL and JSON job spec

Comment (schmichael) [inline nomad/structs/structs.go]: I was going to implement it for non-agent checks in a followup PR (and doc it then), but that seems silly. Fixing.

File: nomad/structs/diff_test.go [hunk 27]
```diff
@@ -1900,20 +1900,22 @@ func TestTaskGroupDiff(t *testing.T) {
 
 func TestTaskDiff(t *testing.T) {
 	cases := []struct {
+		Name       string
```
Comment (dadgar) [inline nomad/structs/diff_test.go]: Bless your soul

**Pull request merged**